File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        left, right = 0, 1      # Buy, Sell
        
        while right < len(prices):
            curr_profit = prices[right] - prices[left]
            max_profit = max(max_profit, curr_profit)

            if prices[left] > prices[right]:
                left = right
            right += 1
            
        return max_profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.ts
// set a pointer to buy (index) and pointer to sell (index)
// use for loop to iterate through the input prices array and move the pointers buy and sell day
// calculate the profit by subtracting sell and buy, set that as max profit
// return max profit or 0 if no profit

function maxProfit(prices: number[]): number {
    let maxProfit: number = 0
    let buyDay: number = 0
    let sellDay: number = 1

    while (sellDay < prices.length) {
        const profit: number = prices[sellDay] - prices[buyDay]
        maxProfit = Math.max(profit, maxProfit)

        if (prices[sellDay] < prices[buyDay]) {
            buyDay = sellDay
        }
        sellDay++
    }
    
    return maxProfit
};